README.md
# Calcolo Numerico

*Esercizi di calcolo numerico (teoria da libro + laboratorio)*


## Install
1. just provide yourself with a working installation of [Matlab Mathworks](https://it.mathworks.com/products/matlab.html) or [Gnu/Octave](https://www.gnu.org/software/octave/)
2. create a script and run [some](/mrz-3-edition) functions .. or use the [tester](/mrz-3-edition/tester.m)

## Questions and issues
The [Github issue tracker](https://github.com/FIUP/calcolo-numerico/issues) is **only** for bug reports and feature requests. Anything else, such as questions for help, should be posted as `pull request` with detailed motivation.

## Important note
Our university provide us a full license of MATLAB for free. You can find more information here (In italian)
[MATLAB | CSIA](https://www.csia.unipd.it/servizi/servizi-utenti-istituzionali/contratti-software-e-licenze/matlab)

## License
[Apache License](http://www.apache.org/licenses/LICENSE-2.0) Version 2.0, January 2004


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

# TODO:
## lab 1: es 4

## 0.4.6- 2017-06-02
### Added
- lab tests
- lab 0: exercises
- lab 1: parabola solver

## 0.4.5 - 2017-05-29
### Added
- exercise 7.3, 7.4, 7.5
- classical and cyclic jacobi eigenv* finder
- hessenberg tranformer with givens rotations
- householder hessenberg transformer
- qr without shift method
- exercise 5.4

## 0.4.4 - 2017-05-28
### Added
- trapezoidal rule, simpson, romberg integration methods
- exercise 7.1, 7.2 with image

## 0.4.3 - 2017-05-26
### Added
- lower/upper triangular matrix inverter
- gauss inverter
- cholesky linear system solver
- jacobi, Gauss-Seidel, SOR, conjugate gradient, pre-conditioned gradient methods
- exercise 4.7, 4.8, 4.9, 4.10, 4.11, 4.12, 4.13 with image

### Fixed
- && operator in Jacobi function

## 0.4.2 - 2017-05-25
### Refactored
- some functions names

### Added
- exercise 4.4, 4.5, 4.6
- gauss doolittle, tridiagonal Thomas methods

## 0.4.1 - 2017-05-24
### Fixed
- L creation on Gauss solvers

### Added
- exercise 4.3

## 0.4 - 2017-05-03

### Added
- gaussPivotingWithPermutationSolver, 4.2 exercises

## 0.3.9 - 2017-05-02

### Added
- gaussLuSolver, gaussPivotingSolver, gaussCroutLuFactor

## 0.3.8 - 2017-04-14

### Added
- upper/lower triangular matrix solver functions

## 0.3.7 - 2017-04-14

### Added
- exercise 6.8, 6.10, 6.9 with images
- least square method
- squared error function
- (prepared file for) evaluation of polynomial in multiple points
- cramer solver

## 0.3.6 - 2017-04-07

### Added
- exercise6.7 with image, exercise6.4

## 0.3.5 - 2017-04-06

### Added
- chebyshevPolynomia builder, chebyshevPolynomiaZeros finder

## 0.3.4 - 2017-04-05

### Added
- forwardFiniteDifferencesTable and binomial way
- exercise 6.6

## 0.3.3 - 2017-04-03

### Added
- dividedDifferencesTable, exercise 6.1, 6.5 with image, newtonInterpolation

## 0.3.2 - 2017-04-02

### Added
- lagrange interpolation

## 0.3.1 - 2017-03-30

### Added
- exercise 3.6, 3.7, 3.8 with image

## 0.3 - 2017-03-29

### Added
- aitken (coding with plaster, sorry for mi-spellings), newtonPlotter with image, steffensenPlotter with image, bisectionVSAnomymous with image
- `utils` module

## 0.2.9 - 2017-03-26

### Added
- help and docs on functions, multiplicityEvaluation

## 0.2.8 - 2017-03-23

### Added
- convergenceEvaluation

### Fixed
- halley method iteration due to book error

## 0.2.7 - 2017-03-22

### Added
- secantVSNewton plot and image

### Fixed
- not overwriting new x value bug in secant method
- improved secant method by pre-allocating f(x) values
- README link to chapters

## 0.2.6 - 2017-03-22

### Added
- algs: fixedPoint, newton, multipleRootsNewton, halley
- es: whyBisectLikeThis, secant

### Fixed
- newton, multipleRootsNewton names

## 0.2.5 - 2017-03-17

### Added
- parabolaSolverComplexVersion

## 0.2.4 - 2017-03-16

### Added
- maxMinFinder, longSummationSolver, squaredMultiplications and squaredMultiplicationsTester, polynomialEvaluationInPoint, parabolaSolverTester

### Fixed
- polynomialHornerEvaluationInPoint: now user can omit the polynomial degree

## 0.2.3 - 2017-03-16

### Added
- epsEstimation, baseComputation, parabolaSolver
- raw II chapter algs and exercises

### Refactored
- chapters

## 0.2.2 - 2017-03-15

### Added
- polynomialDerivativeEvaluationInPoint

## 0.2.1 - 2017-03-15

### Added
- fromBase10FractionalPartConversion

## 0.2 - 2017-03-09

### Added
- polynomialEvaluationInPoint
- toBase10FractionalPartHornerConversion

## 0.1 - 2017-03-08

### Added
- fromBase10Conversion
- toBase10HornerConversion
- guidelines to use tester.m

## 0.0 - 2017-03-08

### Created
- repository


book/tester.m
% Some examples of implementations of the numerical anaysis algorithms and
% exercises in book by prof. Michela Redivo Zaglia
% To run scripts in this repository go to Home > Environment > Set Path >
% Add with subfolders > Select this folder

% Copyright 2017 Stefano Fogarollo
%
% Licensed under the Apache License, Version 2.0 (the "License");
% you may not use this file except in compliance with the License.
% You may obtain a copy of the License at
%
% http://www.apache.org/licenses/LICENSE-2.0
%
% Unless required by applicable law or agreed to in writing, software
% distributed under the License is distributed on an "AS IS" BASIS,
% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
% See the License for the specific language governing permissions and
% limitations under the License.

close all; clc;  % clear memory

%% chapter 1
% algorithms
fromBase10Convert(4, 2);  % 100
polynomialHornerEvaluationInPoint([2, 1, 1], 10);  % 211
toBase10HornerConversion([15, 0, 1], 16);  % 3841

% exercises
toBase10FractionalPartHornerConversion([0, 0, 1], 2);  % 0.1250
fromBase10FractionalPartConversion(1/7, 4, 8);  % [0, 2, 1, 0, 2, 1, 0, 2, 1]
polynomialDerivativeEvaluationInPoint([3, 4, 8, 2, 9], 7);  % 4818

%% chapter 2
% algorithms
epsEstimation();  % 2.2204e-16 (Matlab R2016b in Linux 64bit)
baseComputation();  % 1 (Matlab R2016b in Linux 64bit)
parabolaSolver(1, 4, 1);  % -3.7321, -0.2679

% exercises
maxMinFinder(1.9, 2.0);  % 1.900000000000001, 1.999999999999999 (Matlab R2016b in Linux 64bit)
longSummationSolver([19 * 10^3, 0.4, 5 * 10 ^ (-2), 4.69999999]);  % 1.900514999999000e+04
squaredMultiplicationsTester(2, 3);
    % [ 2,2.22044604925031e-16,2.22044604925031e-16,16;
    % 3,1.11022302462516e-16,1.11022302462516e-16,16 ]
hornerVSAll([10^18, 10 ^ 1, 10^-9], 10^23);  % 9.999999999999999e+63, 9.999999999999999e+63
parabolaSolverSmartVersion(1, 4, 1);  % -3.7321, -0.2679
% parabolaSolverTester(10 ^(-3), 0.8, -6 * 10 ^ (-5), -1 * 10 ^ (-5));  % image 2.2
parabolaSolverComplexVersion(9, 1, 1);  % -0.0556 - 0.3287i, -0.0556 + 0.3287i

%% chapter 3
% algorithms
fixedPoint(@sin, 0.5, 10^(-8), 100);  % 0.1627, 100
newton(@sin, @cos, 1, 10^(-8), 100);  % 0, 5, false
multipleRootsNewton(@(x) x.^2, @(x) 2 * x, 3, 10^(-16), 100, 1);  % 5.551115123125783e-17, 54
halley(@(x) -5 * x.^4 + 6, @(x) -20 * x.^3, @(x) -60 * x.^2, 1, 10^(-16), 100);  % 1.046635139392106, 4
convergenceEvaluation(10 ^ (-16), 100);  % 1.0e+10, 2.0, 1.34e+5, 73
multiplicityEvaluation(@(x) x.^2, @(x) 2 * x, 1, 10^(-16), 100);  % 5.551115123125783e-17, 0.5, 2, 54
% laguerre(); TODO
% bairstow(); TODO

% exercises
whyBisectLikeThis(0.982, 0.984);  % 0.983000000000000, 0.983000000000000
% secantVSNewton(10^(-10), 10^(-10), 2 * 10^(-10), 100);  % image 3.17
% newtonPlotter(2);  % image 3.20
% steffensenPlotter(1);  % exercise 3.4
% bisectionVSAnomymous(1.5);  % exercise 3.5
% exercise3_6();  % exercise 3.6
% exercise3_7();  % exercise 3.7
% exercise3_8();  % exercise 3.8

%% chapter 4
% algorithms
upTriangSolve([
    1 3;
    0 4
], [3 5]');  % [-0.75 1.25]'
lowTriangSolve([
    1 0;
    3 4
], [3 5]');  % [3 -1]'
gaussLu([
    1 7 8;
    1 3 1
    5 4 3
], [9 8 2]');
upTriangSolve(U, y);  % [-1.4348 3.8261 -2.0435]'
gaussPivoting([
    1 1 1;
    0 2 5;
    2 5 -1
], [6 -4 27]');
upTriangSolve(U, y);  % [5 3 -2]'
gaussLuCrout([4 2 1; 9 8 7; 2 2 9]);  % [4 0 0; 9 3.5 0; 2 1 7.1429], [1 0.5 0.25; 0 1 1.3571; 0 0 1]
cholesky([74 31 100; 31 54 89; 100 89 226]);  % [8.6023 0 0; 3.6037 6.4042 0; 11.6248 7.3558 6.0627]

% exercises
gaussPivotingPerm([
    2 1 3;
    4 3 10;
    2 4 17
    ], [11 28 31]');  % example 4.3 in book with exercise 4.1 function
exercise4_2();  % [0.1546 1.2336 1.9710 1.4386 1.6432]'
% exercise4_3();  % [7.5383 31.3009 -6.4753 -17.7043]'
exercise4_4();  % [3.8 -6.2 2.8 -2.7 8.7 0.85 16.15]'
exercise4_5();  % [-4 60 -180 140]'
exercise4_6();  % [1 1 1 1 1]'
gaussInvert([
    5 -1 3 2 -4;
    -3 7 -1 1 3;
    4 2 -3 8 3;
    3 -3 6 -1 2;
    1 -4 2 -3 7;
]);  % exercise 4.7
choleskySolve([
    5 7 6 5;
    7 10 8 7;
    6 8 10 9;
    5 7 9 10;
], [23 32 33 31]');  % exercise 4.8
% exercise4_9();  % 1.0e-14 * ...
% exercise4_10();  % see image
% exercise4_11();  % see image
% exercise4_12();  % see image
gradConjugate([
    60 -31 26 33 -28;
    -31 79 -42 36 -15;
    26 -42 59 -31 20;
    33 36 -31 79 -52;
    -28 -15 20 -52 87;
], [60 27 32 65 12]', zeros(5, 1), 1e-10, 20);  % exercise 4.13, [1 1 1 1 1]'

%% chapter 5
% algorithms
jacobiEigen([
    14 32 50;
    32 77 122;
    50 122 194
], 1e-8, 10);  % same as eig(...)
jacobiEigenCycle([
    14 32 50;
    32 77 122;
    50 122 194
], 1e-8, 10);  % same as eig(...)
givensTransform([
    14 32 50;
    32 77 122;
    50 122 194
]);  % same as hess(...)
householderTransform([
    14 32 50;
    32 77 122;
    50 122 194
]);  % same as hess(...)
qrNoShift([
    0.6026 -0.6896 0;
    -0.6896 90.3974 -131.8484;
    0 -131.8484 194.0000
], 1e-8, 10)  % same as eig(...)

% exercises
exercise5_4 ();  % [1 2 3 4]

%% chapter 6
% algorithms
% lagrangeApproximate([2.92 3.75 -5.2], [8.24 16.32 2.41], [1.01 1.54 -4.24]);  % [-5.0812 -2.1216 -3.8258]
% nevilleAitkenApproxiate();  % TODO
% dividedDifferencesTable([3 1 5 6], [1 -3 2 4]);  % [1.0000 2.0000 -0.3750 0.1750; -3.0000 1.2500 0.1500 0; 2.0000 2.0000 0 0; 4.0000 0 0 0]
% chebyshevPolynomia(3);  % [4 0 -3 0]
% chebyshevPolynomiaZeros(4);  % [-0.9239 -0.3827 0.3827 0.9239]

% exercises
% exercise6_1();  % exercise 6.1
exercise6_4(linspace(2, 2.4, 5), sqrt(linspace(2, 2.4, 5)));  % [1.414213562373095 0.349241122458488 -0.041104472283004 0.009243006108162 -0.002486810899427], [1.414213562373095 0.349241122458488 -0.041104472283004 0.009243006108147 -0.002486810899327]
% exercise6_5();  % exercise 6.5
% exercise6_6();  % exercise 6.6
% exercise6_7();  % exercise 6.7
% exercise6_8();  % exercise 6.8
% exercise6_9();  % exercise 6.9
% exercise6_10(); % exercise 6.10

%% chapter 7
% algorithms
trapezoidalRule(@(x) x^2, 0, 4, 100);  % 21.33
simpson(@(x) x^3, 0, 4, 100);  % 64
romberg(@(x) x^4, 0, 4, 100, 5);  % 204.8

% exercises
% exercise7_1();  % see image
% exercise7_2();  % see image
% exercise7_3(); % exercise 7.3
% exercise7_4(); % exercise 7.4
% exercise7_5(); % exercise 7.5

%% chapter 8
% algorithms
% exercises